tao_compiler/mlir/disc/transforms/disc_memref_gvn.cc

```diff
@@ -0,0 +1,149 @@
+#include "tensorflow/compiler/mlir/disc/disc_util.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+
+namespace mlir {
+namespace disc_ral {
+
+namespace {
+
+struct LoadScalarSimplifier : public OpRewritePattern<memref::LoadOp> {
+ public:
+  LoadScalarSimplifier(MLIRContext* context, DominanceInfo* dominance_info)
+      : OpRewritePattern<memref::LoadOp>::OpRewritePattern(context),
+        dominance_info_(dominance_info) {}
+
+  LogicalResult matchAndRewrite(memref::LoadOp op,
+                                PatternRewriter& rewriter) const override;
+
+ private:
+  bool extractScalarIndex(ValueRange indices, int64_t& index) const;
+  DominanceInfo* dominance_info_;
+};
+
+bool LoadScalarSimplifier::extractScalarIndex(ValueRange indices,
+                                              int64_t& index) const {
+  if (indices.size() != 1) {
+    return false;
+  }
+  auto op = indices[0].getDefiningOp();
+  if (auto constOp = dyn_cast_or_null<arith::ConstantIndexOp>(op)) {
+    index = constOp.value();
+    return true;
+  }
+  return false;
+}
+
+// Replace the pattern of:
+//   memref.store %value, %memref_x[%index_x] : memref<?xtype>
+//   %a = memref.load %memref_x[%index_x] : memref<?xtype>
+//   Use of `%a`
+// with:
+//   memref.store %value, %memref_x[%index_x] : memref<?xtype>
+//   Use of `%value`
+// The store op could be eliminated in other passes (e.g., DCE).
+// Note: we only deal with memrefs on CPU in this pass currently.
+LogicalResult LoadScalarSimplifier::matchAndRewrite(
+    memref::LoadOp op, PatternRewriter& rewriter) const {
+  Location loc = op.getLoc();
+  int64_t load_index;
+  if (!extractScalarIndex(op.getIndices(), load_index)) {
+    return failure();
+  }
+
+  Value loadMemRef = op.getMemRef();
+  if (placement_utils::isGpuMemRef(loadMemRef)) {
+    return failure();
+  }
+  DenseSet<Operation*> writable_ops;
+  for (Operation* user : loadMemRef.getUsers()) {
+    if (!dominance_info_->dominates(user, op)) {
+      continue;
+    }
+    if (!isa<memref::StoreOp>(user)) {
+      if (IsOpWriteValue(user, loadMemRef)) {
+        writable_ops.insert(user);
+      }
+      continue;
+    }
+    // Load and store should work on the same index.
+    auto store = dyn_cast<memref::StoreOp>(user);
+    int64_t store_index;
+    if (!extractScalarIndex(store.getIndices(), store_index)) {
+      continue;
+    }
+    if (load_index == store_index) {
+      auto value = store.value();
+      writable_ops.insert(user);
+    }
+  }
+  SmallVector<Operation*> write_frontier;
+  for (auto curr : writable_ops) {
+    bool dominant_other = false;
+    for (auto other : writable_ops) {
+      if (!dominance_info_->dominates(curr, other)) {
+        dominant_other = true;
+        break;
+      }
+    }
+    if (!dominant_other) {
+      write_frontier.push_back(curr);
+    }
+  }
+  // Make sure that there is only one deterministic dominant store.
+  if ((write_frontier.size() == 1)) {
```

Comment:
Some corner case：

```
// b not dominates c while may affects the value of c.
memref.store %val1, %memref_x[%index_x] : memref<?xtype> // a
if (%pred) {
  memref.store %val2, %memref_x[%index_x] : memref<?xtype> // b
}
%0 = memref.load %memref_x[%index_x] : memref<?xtype> // c
```

I wonder whether it’s enough for our use case  (only try to optimize shape computation with static shape) if we just do some conservative analysis （e.g. a buffer can only have one store user）.

Moreover, memref alias (generated by cast/view ops) should be taken into consideration as well. Again, general cases is complicated while some  conservative analysis (only do such optimization when there are no alias)  may be enough for our use cases.

Reply:
Good comment! A better implementation is on going.

Reply:
Updated. Please help to check whether some corner cases are missed.